### PR TITLE
Mention the tutorial of `003_efficient_optimization_algorithms.py` in the `Trial` API page

### LINKED ISSUE #2940: Mention each tutorial page in API pages at least once
[Optuna's tutorial](https://optuna.readthedocs.io/en/stable/tutorial/index.html) is a good source to understand Optuna's functionality with concrete examples. However, some tutorial pages might not be mentioned in corresponding Optuna's API pages. 

## Description

Please add a `note section` or `see also section` to mention the following tutorial pages.

- ~[ ] [Lightweight, versatile, and platform agnostic architecture](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/001_first.html)~
    - ~this is is a little bit conceptual page, so it might not be necessary to be linked.~ 
- [x] [Pythonic Search Space](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/002_configurations.html)
- [x] [Efficient Optimization Algorithms](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/003_efficient_optimization_algorithms.html)
- [x] [Easy Parallelization](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/004_distributed.html)
- [x] [Quick Visualization for Hyperparameter Optimization Analysis](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/005_visualization.html) 
  - linked from https://optuna.readthedocs.io/en/stable/reference/visualization/index.html
- [x] [Saving/Resuming Study with RDB Backend](https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/001_rdb.html)
  - linked from https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.create_study.html#optuna.study.create_study
- [x] [Multi-objective Optimization with Optuna] (https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/002_multi_objective.html) https://github.com/optuna/optuna/pull/3339
- [x] [User Attributes](https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/003_attributes.html)
- [x] [Command-Line Interface](https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/004_cli.html)
  - linked from https://optuna.readthedocs.io/en/stable/reference/index.html
- [x] [User-Defined Sampler](https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/005_user_defined_sampler.html)
  - linked from https://optuna.readthedocs.io/en/stable/reference/samplers.html
- [x] [User-Defined Pruner](https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/006_user_defined_pruner.html)
  - linked from https://optuna.readthedocs.io/en/stable/reference/pruners.html
- [x] [Callback for Study.optimize](https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/007_optuna_callback.html)
  - linked from `callbacks`'s section in https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.optimize
- [ ] [Specify Hyperparameters Manually](https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/008_specify_params.html)
  - [x] [the first section](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.optimize) can be linked from `Study.enqueue_trial` 
  - [ ] [the second section](https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/008_specify_params.html#second-scenario-have-optuna-utilize-already-evaluated-hyperparameters) can be linked from `Study.add_trial`　(see https://github.com/optuna/optuna/pull/3346)
- [x] [Ask-and-Tell Interface](https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/009_ask_and_tell.html)
- [x] [Re-use the best values](https://optuna.readthedocs.io/en/stable/tutorial/20_recipes/010_reuse_best_trial.html) https://github.com/optuna/optuna/pull/3396

### PR DESCRIPTION
**Pull request by makkimaki**
## Motivation

Partially fixed #2940. I think this PR completes ` Efficient Optimization Algorithms` item in the check list.

## Description of the changes

To `optuna/docs/source/reference/pruners.rst` & `optuna/docs/source/reference/samplers.rst`, seealso link of tutorial is addded respectively.

### CODE REVIEW

**Comment by HideakiImamura** [comment]
How about adding not only `optuna.samplers` but also `optuna.pruners`?

**Comment by makkimaki** [comment]
> How about adding not only `optuna.samplers` but also `optuna.pruners`?

ok. I should have known the base class relationships more in details. Thanks.

**Comment by HideakiImamura** [comment]
I think `optuna.pruners` should be added around line 56-73.

**Comment by makkimaki** [comment]
> I think `optuna.pruners` should be added around line 56-73 since this tutorial explains samplers and pruners.

I add the new one. How about it?

File: tutorial/10_key_features/003_efficient_optimization_algorithms.py [hunk 1]
```diff
@@ -72,7 +75,8 @@
 # though basically it is outperformed by :class:`optuna.pruners.SuccessiveHalvingPruner` and
 # :class:`optuna.pruners.HyperbandPruner` as in `this benchmark result <https://github.com/optuna/optuna/wiki/Benchmarks-with-Kurobako>`_.
 #
-#
+#.. seealso::
```
Comment (makkimaki) [inline tutorial/10_key_features/003_efficient_optimization_algorithms.py:78]: I have little confidence in these following lines. Is it ok `#` exists in front of the new comments..?

**Comment by makkimaki** [comment]
@HideakiImamura How about this change?

**Comment by HideakiImamura** [comment]
@nzw0301 Could you review this PR?

**Comment by nzw0301** [comment]
Any update?

**Comment by HideakiImamura** [comment]
@makkimaki @nzw0301 How about adding a `seealso` link of `003_efficient_optimization_algorithms.py` in [here](https://optuna.readthedocs.io/en/stable/reference/samplers.html) and [here](https://optuna.readthedocs.io/en/stable/reference/pruners.html)?

**Comment by nzw0301** [comment]
I suppose @HideakiImamura's [comment](https://github.com/optuna/optuna/pull/3068#issuecomment-962758368) refers
https://github.com/optuna/optuna/blob/master/docs/source/reference/pruners.rst and https://github.com/optuna/optuna/blob/master/docs/source/reference/samplers.rst not the all classes on these pages.

**Comment by makkimaki** [comment]
It completely slipped my mind. I was used to copying the link in the same way as `Pythonic Search Space` tutorial. I'm going to modify. Do I supposed to add a `seealso` link to those 2 pages you mentioned? @nzw0301

**Comment by nzw0301** [comment]
@makkimaki no worries. 

More specifically, for https://github.com/optuna/optuna/blob/master/docs/source/reference/pruners.rst, the content will become as follows by this PR:

```rst
.. module:: optuna.pruners

optuna.pruners
==============

The :mod:`~optuna.pruners` module defines a :class:`~optuna.pruners.BasePruner` class characterized by an abstract :meth:`~optuna.pruners.BasePruner.prune` method, which, for a given trial and its associated study, returns a boolean value representing whether the trial should be pruned. This determination is made based on stored intermediate values of the objective function, as previously reported for the trial using :meth:`optuna.trial.Trial.report`. The remaining classes in this module represent child classes, inheriting from :class:`~optuna.pruners.BasePruner`, which implement different pruning strategies.

.. seealso::
    :ref:`pruning` tutorial explains the concept of the pruner classes and a minimal example.

.. autosummary::
   :toctree: generated/
   :nosignatures:

   optuna.pruners.BasePruner
   optuna.pruners.MedianPruner
   optuna.pruners.NopPruner
   optuna.pruners.PatientPruner
   optuna.pruners.PercentilePruner
   optuna.pruners.SuccessiveHalvingPruner
   optuna.pruners.HyperbandPruner
   optuna.pruners.ThresholdPruner
```

So, this PR adds the following lines

```rst
.. seealso::
    :ref:`pruning` tutorial explains the concept of the pruner classes and a minimal example.
```

to [the file](https://github.com/optuna/optuna/blob/master/docs/source/reference/pruners.rst).

--- 

Similarly, 

More specifically, for https://github.com/optuna/optuna/blob/master/docs/source/reference/samplers.rst, the content will become as follows by this PR:


```rst
.. module:: optuna.samplers

optuna.samplers
===============

The :mod:`~optuna.samplers` module defines a base class for parameter sampling as described extensively in :class:`~optuna.samplers.BaseSampler`. The remaining classes in this module represent child classes, deriving from :class:`~optuna.samplers.BaseSampler`, which implement different sampling strategies.

.. seealso::
    :ref:`pruning` tutorial explains the overview of the sampler classes.

.. autosummary::
   :toctree: generated/
   :nosignatures:

   optuna.samplers.BaseSampler
   optuna.samplers.GridSampler
   optuna.samplers.RandomSampler
   optuna.samplers.TPESampler
   optuna.samplers.CmaEsSampler
   optuna.samplers.PartialFixedSampler
   optuna.samplers.NSGAIISampler
   optuna.samplers.MOTPESampler
   optuna.samplers.IntersectionSearchSpace
   optuna.samplers.intersection_search_space
```


Likewise, this PR adds the following lines

```rst
.. seealso::
    :ref:`pruning` tutorial explains the overview of the sampler classes.
```

to [the file](https://github.com/optuna/optuna/blob/master/docs/source/reference/samplers.rst).

**Comment by nzw0301** [comment]
@HideakiImamura  Is my understanding correct?

**Comment by HideakiImamura** [comment]
@nzw0301 Yes, it is the same in my mind. 

@makkimaki Could you fix them as nzw0301 suggested?

**Comment by HideakiImamura** [comment]
@makkimaki The CI error in the master branch is fixed in #3103. Could you merge the master branch to pass the CI?

**Review by nzw0301 — APPROVED** [review]

**Review by HideakiImamura — APPROVED** [review]
LGTM.

**Pull request merged**